### ISSUE TITLE
gcc/g++ compiled 32 bits dll doesn't work

### ISSUE THREAD

**Issue by lifenjoiner** [ScintillaOrg/lexilla#10]
The reason is that the exported stdcall function names contains `@n` suffixes. `GetProcAddress` on the origin function name does not work.
`--kill-at`: https://sourceware.org/binutils/docs/ld/Options.html

Patch:
[Strip-stdcall-suffixes-from-exported-symbols-generated-by-MinGW-w64.zip](https://github.com/ScintillaOrg/lexilla/files/6509333/Strip-stdcall-suffixes-from-exported-symbols-generated-by-MinGW-w64.zip)

BTW:
[Scintilla](https://sourceforge.net/p/scintilla/code/ci/e245ae2fb15b60ddbd60c64c45f6c1df382726e3/tree/win32/makefile#l47) also needs it.

**Comment by nyamatongwe** [comment]
gcc/g++ compiled Lexilla.dll does work. Its one of the most tested paths.
~~~text
C:\u\hg\lexilla\bin>dumpbin /exports lexilla.dll
Microsoft (R) COFF/PE Dumper Version 14.28.29915.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file lexilla.dll

File Type: DLL

  Section contains the following exports for lexilla.dll

    00000000 characteristics
    60A58900 time date stamp Thu May 20 07:54:08 2021
        0.00 version
           1 ordinal base
           8 number of functions
           8 number of names

    ordinal hint RVA      name

          1    0 00001639 CreateLexer
          2    1 00001581 GetLexerCount
          3    2 00001618 GetLexerFactory
          4    3 000015A1 GetLexerName
          5    4 00001510 GetLibraryPropertyNames
          6    5 00001519 GetNameSpace
          7    6 0000169E LexerNameFromID
          8    7 00001518 SetLibraryProperty
~~~

**Comment by lifenjoiner** [comment]
That's strange. Can you run `TestLexers` successfully? I can't.

Here is my screenshot:
![TestLexers-Err](https://user-images.githubusercontent.com/19585474/118905874-2dfc5a80-b94f-11eb-908e-c86c23706338.png)

Due to `GetProcAddress` failure with mismatched function names:
![stdcall-gcc-exports](https://user-images.githubusercontent.com/19585474/118904555-b1687c80-b94c-11eb-8b02-58d96e052337.png)

Ref:
https://www.willus.com/mingw/yongweiwu_stdcall.html

**Comment by lifenjoiner** [comment]
After testing **x64** builds, I found that it doesn't have the `@n` suffixes.
So, striping the `@n` suffixes explicitly is required by **i686** GCC.

**Comment by rdipardo** [comment]
Not reproducible:
```
C:\Users\Rob\git\scintilla-contrib\lexilla\bin>dumpbin /exports lexilla.dll
Microsoft (R) COFF/PE Dumper Version 14.28.29914.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file lexilla.dll

File Type: DLL

  Section contains the following exports for lexilla.dll

    00000000 characteristics
    FFFFFFFF time date stamp
        0.00 version
           1 ordinal base
           8 number of functions
           8 number of names

    ordinal hint RVA      name

          1    0 00001705 CreateLexer = _CreateLexer@4
          2    1 000016A6 GetLexerCount = _GetLexerCount@0
          3    2 000016EE GetLexerFactory = _GetLexerFactory@4
          4    3 000016BA GetLexerName = _GetLexerName@12
          5    4 000017B4 GetLibraryPropertyNames = _GetLibraryPropertyNames@0
          6    5 000017BD GetNameSpace = _GetNameSpace@0
          8    6 000017BA SetLibraryProperty = _SetLibraryProperty@8
          7    7 0000177F _LexerNameFromID@4 = _LexerNameFromID@4

  Summary

        4000 .data
       1A000 .rdata
        5000 .reloc
        1000 .rsrc
       7A000 .text

C:\Users\Rob\git\scintilla-contrib\lexilla\bin>cd ../test

C:\Users\Rob\git\scintilla-contrib\lexilla\test>nmake -f testlexers.mak test

Microsoft (R) Program Maintenance Utility Version 14.28.29914.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        TestLexers.exe
Found Lexilla at C:\Users\Rob\git\scintilla-contrib\lexilla
Lexing batch\x.bat
Lexing cpp\Bug2245.cxx
Lexing cpp\x.cxx
Lexing d\x.d
Lexing errorlist\AllStyles.err
Lexing fsharp\x.fs
Lexing hypertext\apostophe.php
Lexing hypertext\Bug2207.html
Lexing hypertext\x.asp
Lexing hypertext\x.html
Lexing hypertext\x.php
Lexing latex\AllStyles.tex
Lexing latex\Feature1358.tex
Lexing lua\Bug2205.lua
Lexing lua\x.lua
Lexing makefile\x.mak
Lexing markdown\AllStyles.md
Lexing markdown\Bug2235.md
Lexing mmixal\AllStyles.mms
Lexing mmixal\references.mms
Lexing mmixal\x.mms
Lexing nim\x.nim
Lexing perl\perl-test-5220delta.pl
Lexing perl\perl-test-sub-prototypes.pl
Lexing perl\x.pl
Lexing python\AllStyles.py
Lexing python\x.py
Lexing raku\x.p6
Lexing ruby\x.rb
Lexing tcl\x.tcl
Lexing vb\x.vb
```

**Comment by rdipardo** [comment]
> Here is my screenshot:
> ![TestLexers-Err](https://user-images.githubusercontent.com/19585474/118905874-2dfc5a80-b94f-11eb-908e-c86c23706338.png)
> 

Check the file endings of the `*.properties` files in the `test/example` directory:

         git ls-files --eol test/examples/*.properties

I've noticed that `TestLexers` will fail to parse `*.properties` files when they have CRLF line endings on a Unix platform such as MYSYS2 (as I see in the title bar of your capture).  This could happen if you cloned the source with a Windows git client, then compiled in a MYSYS2 session. Here's a demonstration using WSL:

```
rob@AcerNotebook:/mnt/c/Users/Rob/sourcegit/scintilla-contrib/lexilla (master=)
$ uname -a
Linux AcerNotebook 4.4.0-19041-Microsoft #488-Microsoft Mon Sep 01 13:43:00 PST 2020 x86_64 x86_64 x86_64 GNU/Linux

rob@AcerNotebook:/mnt/c/Users/Rob/sourcegit/scintilla-contrib/lexilla (master=)
$ git ls-files --eol test/examples/*.properties
i/lf    w/crlf  attr/text               test/examples/batch/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/cpp/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/d/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/errorlist/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/fsharp/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/hypertext/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/julia/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/latex/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/lua/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/makefile/SciTE.properties
i/-text w/-text attr/text               test/examples/markdown/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/mmixal/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/nim/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/perl/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/python/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/raku/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/ruby/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/tcl/SciTE.properties
i/lf    w/crlf  attr/text               test/examples/vb/SciTE.properties

rob@AcerNotebook:/mnt/c/Users/Rob/sourcegit/scintilla-contrib/lexilla (master=)
$ cd src && make
g++ -DNDEBUG -I ../include -I ../../scintilla/include -I ../src -I ../lexlib -fvisibility=hidden --std=c++17 -fPIC -Os -Wpedantic -Wall -Wextra   -c Lexilla.cxx -o Lexilla.o
[ . . . ]
ranlib ../bin/liblexilla.a

rob@AcerNotebook:/mnt/c/Users/Rob/source/git/scintilla-contrib/lexilla/src (master=)
$ cd ../test && make test
g++ -DNDEBUG -I ../../scintilla/include -I ../include -I ../access --std=c++2a -Os -Wpedantic -Wall -Wextra   -c TestLexers.cxx -o TestLexers.o
g++ -DNDEBUG -I ../../scintilla/include -I ../include -I ../access --std=c++2a -Os -Wpedantic -Wall -Wextra   -c TestDocument.cxx -o TestDocument.o
g++ -DNDEBUG -I ../../scintilla/include -I ../include -I ../access --std=c++2a -Os -Wpedantic -Wall -Wextra   -c ../access/LexillaAccess.cxx -o LexillaAccess.o
g++ --std=c++2a -Os -Wpedantic -Wall -Wextra   TestLexers.o TestDocument.o LexillaAccess.o -ldl  -o TestLexers
./TestLexers
Found Lexilla at /mnt/c/Users/Rob/source/git/scintilla-contrib/lexilla
Lexing batch/x.bat

/mnt/c/Users/Rob/source/git/scintilla-contrib/lexilla/test/examples/batch/x.bat:1: has no lexer

Lexing cpp/Bug2245.cxx

/mnt/c/Users/Rob/source/git/scintilla-contrib/lexilla/test/examples/cpp/Bug2245.cxx:1: has no lexer

Lexing cpp/x.cxx

/mnt/c/Users/Rob/source/git/scintilla-contrib/lexilla/test/examples/cpp/x.cxx:1: has no lexer

Lexing d/x.d

/mnt/c/Users/Rob/source/git/scintilla-contrib/lexilla/test/examples/d/x.d:1: has no lexer
[ . . . ]
make: *** [makefile:55: test] Error 1
```
As a guess, the fault could lie with `TestLexers.cxx` calling `std:getline` on the `*.properties` files without taking the system's line delimiter into account: 

https://github.com/ScintillaOrg/lexilla/blob/6ee6b086a1920fdafe3af41eecbd865d52162ec7/test/TestLexers.cxx#L87-L92

**Comment by nyamatongwe** [comment]
Added a chomp after the getline with 3377db1832265dbb414f207192a862d3d878eb58.

**Comment by lifenjoiner** [comment]
@rdipardo Thanks for you help.
1. I think you also generated export function names with `@n` suffixes.
2. Here you use `nmake`. Are you using GCC? Which version? I used (`MSYS2` + )`i686-GCC` (10.2.0 as the screenshot shows).
`1    0 00001705 CreateLexer = _CreateLexer@4` shows it exported both  `_CreateLexer@4` and `CreateLexer`. You can also observe it with VS shipped `depends.exe` or other PE tools.
3. I debugged the `TestLexers.exe` with `x32dbg`, and found `GetProcAddress` fails with return value 0 (NULL). And with `--kill-at` linking option, it works well.
https://github.com/ScintillaOrg/lexilla/blob/3377db1832265dbb414f207192a862d3d878eb58/access/LexillaAccess.cxx#L172-L181

And please refer to `ld` [Options Specific to i386 PE Targets](https://sourceware.org/binutils/docs/ld/Options.html#Options-Specific-to-i386-PE-Targets). It says "`--kill-at`  is specific to the i386 PE targeted port of the linker". It could be a reason with this option there.

BTW:
I use Git for Windows which does the `\r\n` and `\r` conversion automatically.
I'm using Windows 10 20H2.

**Comment by lifenjoiner** [comment]
Actually, `CALLING_CONVENTION` is not necessary for Windows.
https://github.com/ScintillaOrg/lexilla/blob/3377db1832265dbb414f207192a862d3d878eb58/src/Lexilla.cxx#L336
https://github.com/ScintillaOrg/lexilla/blob/3377db1832265dbb414f207192a862d3d878eb58/src/Lexilla.cxx#L16

Different [types of functions](https://docs.microsoft.com/en-us/cpp/cpp/argument-passing-and-naming-conventions) calling are recognized/converted while compiling/linking with header files declaring.
https://github.com/ScintillaOrg/lexilla/blob/3377db1832265dbb414f207192a862d3d878eb58/include/Lexilla.h#L87
https://github.com/ScintillaOrg/lexilla/blob/3377db1832265dbb414f207192a862d3d878eb58/include/Lexilla.h#L27

The saying "[The __stdcall calling convention is used to call Win32 API functions](https://docs.microsoft.com/en-us/cpp/cpp/stdcall)" is for `WINAPI` which is used by **Windows OS** APIs, not requiring all to adopt it, I think, just default `__cdecl` works well.

**Comment by lifenjoiner** [comment]
Here I added a CI testing file: https://github.com/lifenjoiner/lexilla/tree/CI-Win
As `windir` in `makefile` doesn't work with MSYS2, it just uses `mingw32-make`.

You can see the failure: https://github.com/lifenjoiner/lexilla/runs/2648803500
Only `mingw32-LEXILLA_SHARED` testing fails.

**Comment by lifenjoiner** [comment]
And with https://github.com/lifenjoiner/lexilla/commit/a0a4ccebfc35654292bec409593129b107446173, all works well: https://github.com/lifenjoiner/lexilla/actions/runs/868212808.

**Comment by nyamatongwe** [comment]
> Actually, `CALLING_CONVENTION` is not necessary for Windows.

A different calling convention would work but its necessary to have both the library and caller use the same calling convention so its specified. Since x64 has a common calling convention, you don't see failures there but on x86 calling convention mismatches can cause crashes.

ILexer uses stdcall as stdcall is used for COM and interfaces like ILexer are similar to COM. Then, for consistency, the external lexer interface was stdcall as well. Lexilla extends the external lexer interface and it would be inconsistent for the new functions to use a different calling convention.

**Comment by lifenjoiner** [comment]
Anyway, problems are there: reported, reproduced, proved, and patches provided. Also https://github.com/ScintillaOrg/lexilla/issues/6. I have finished my work.

**Comment by nyamatongwe** [comment]
clang is using Microsoft link which doesn't like `--kill-at` so there is an additional condition in 
9e60005

**Comment by nyamatongwe** [comment]
Closed as change included in 5.0.3 release.
